working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,431 @@
+written in imperative Dart code. These typically take the form of kernel
+transformers or separate build processes like build_runner which run before the
+compilers and are not directly integrated.
+
+Rust has a nice [declarative, rule-based macro system][rust_declarative_macros].
+But, because that rule-based system is limited, Rust also has a second
+[imperative procedural macro system][rust_procedural_macros] for when you need
+more power. In Dart, we would prefer to only have one metaprogramming system, so
+it should be one that is powerful enough to cover all known use cases.
+
+## Challenges
+
+Having said that, allowing the whole Dart language to run at compile-time
+exposes a number of really difficult problems we will have to solve in 2021:
+
+### Scoping and hygiene
+
+A macro may introduce new symbols into an existing scope (where that function is
+used). How do we deal with conflicts here? Would the generated code exist in its
+own scope and not have the ability to leak any new variables into the
+surrounding scope?
+
+Generated code may also need to access variables in at least 3 different scopes —
+the scope where the macro is applied, the scope of the generated code (may be
+same as the former), and the scope where the macro is defined. How does it
+distinguish between these?
+
+### Ordering and staging
+
+In what order are macros run? If they can both generate new APIs and introspect
+on the program, then ordering can be user-visible. Consider a macro that adds a
+field for memoization and another that adds a method to serialize all fields. If
+both are applied to the same class, does the memoized field get serialized?
+
+Do macros run before type checking, after, or interleaved? What is the static
+type context in which a macro runs?
+
+How do macros interact with const canonicalization and evaluation, do they run
+before, after, or interleaved? Can macros read const values *and* generate them?
+
+There are additional considerations if we allow recursive macros, and cycles
+between macro defining libraries.
+
+### Usability
+
+Macros can easily lead to incomprehensible programs if they become overly
+magical, or are a complete black box to the user. In order to demystify macros
+you should:
+
+* Be able to easily navigate to the macro implementation.
+* Be able to visualize in some way the code generated into your program, at
+  development time.
+* Be able to debug generated code (step through it, etc).
+  * Possibly provide an opt out for this, if feasible/desirable.
+* Be able to trace errors that flow through generated code, as well as navigate
+  back to the line in the macro implementation that produced the code.
+* Be able to auto-complete macro generated apis.
+
+### Performance
+
+A Turing-complete programming language that runs in your typechecker opens the
+door to user-code that locks the IDE. How do we ensure that users maintain a
+fast edit refresh cycle when arbitrary Dart code may be running during compilation?
```

Comment:
The halting problem has one solution: timeouts. I recommend displaying a message after a few seconds that says something along the lines of "Macro blabla is taking a long time to execute. Press ^C to abort, O to obtain an Observatory URL with a debugger, S to print a stack trace, and W to get information on what code was being compiled when this macro was found."

Reply:
Yes, something like this will needed. We will otherwise run the risk of people filing issues about slow builds when its just the fault of some very expensive macro :). I am sure that will happen anyways, but we want to mitigate it for sure.

We may want to do it specifically by limiting recursion depth or something like that instead of a raw timeout (to have more consistent results across machines). But it is the same principle.